base/compiler/typeinfer.jl

```diff
@@ -918,6 +918,33 @@ function typeinf_code(interp::AbstractInterpreter, method::Method, @nospecialize
     return code, rt
 end
 
+"""
+    typeinf_ircode(
+        interp::AbstractInterpreter,
+        method::Method,
+        atype,
+        sparams::SimpleVector,
+    ) -> (ir::Union{IRCode,Nothing}, returntype::Type)
+
+Infer a `method` and return an optimized `IRCode` with inferred `returntype` on success.
+"""
+function typeinf_ircode(interp::AbstractInterpreter, method::Method, @nospecialize(atype), sparams::SimpleVector)
+    mi = specialize_method(method, atype, sparams)::MethodInstance
+    ccall(:jl_typeinf_begin, Cvoid, ())
+    result = InferenceResult(mi)
+    frame = InferenceState(result, :no, interp)
+    if !(frame !== nothing && typeinf(interp, frame) && frame.inferred)
+        ccall(:jl_typeinf_end, Cvoid, ())
+        return nothing, Any
+    end
+    opt_params = OptimizationParams(interp)
+    opt = OptimizationState(frame, opt_params, interp)
+    ir = run_passes(opt.src, opt, result)
```

Comment:
Looks good overall! Would adding an option to toggle optimization of this IR be useful?

Reply:
Yeah, it sounds reasonable to have an option to return `IRCode` that hasn't been optimized yet (i.e. that just created from `convert_to_ircode`).

Reply:
I added the `optimize` flag d5a9dc2b5a8537c34e328b80c9b73d55aa076480

```julia
julia> Base.code_ircode(; optimize = false) do f
           identity(f())
           identity(f())
           f()
       end |> only |> first
2 1 ─ %1 = (_2)()::Any                                                                             │
  │        Main.identity(%1)::Any                                                                  │
3 │   %3 = (_2)()::Any                                                                             │
  │        Main.identity(%3)::Any                                                                  │
4 │   %5 = (_2)()::Any                                                                             │
  └──      return %5                                                                               │


julia> Base.code_ircode() do f
           identity(f())
           identity(f())
           f()
       end |> only |> first
2 1 ─      (_2)()::Any                                                                             │
3 │        (_2)()::Any                                                                             │
4 │   %3 = (_2)()::Any                                                                             │
  └──      return %3                                                                               │
```

I created `run_minimal_passes` to run `convert_to_ircode`, `slot2reg`, and `compact!`: https://github.com/JuliaLang/julia/blob/d5a9dc2b5a8537c34e328b80c9b73d55aa076480/base/compiler/optimize.jl#L548

Since most of the transformations on IRCode do not handle slots, I think it's better to run `slot2reg`. But maybe we should skip `compact!` when `optimize = false`? What do you think?

Reply:
I agree with running `compact!` as I found it more useful to run `compact!` when writing a similar tooling for EA.jl (https://github.com/JuliaLang/julia/blob/d6a4417af880d140e4e3702550906c8fc25bdd50/test/compiler/EscapeAnalysis/EAUtils.jl#L223) especially because we will see some "unfixed" states  that make the testing a bit difficult without `compact!`.

Reply:
Yeah, that's why I wanted to include `compact!` even with `optimize = false`. A non-`compact!`-ed IR is a bit tricky to work with and also the printing is misleading sometimes.

`code_ircode` is a "hidden" feature for interactive use so I don't think we need to be picky about getting it "right" at the first shot. So I suggest keeping it as-is (i.e., include `compact!`) for now unless someone with strong opinions comes before merging this.